content/2021-06-07-quarkus-func-payload-mapper.md

```diff
@@ -0,0 +1,308 @@
+  Ce-Type:io.drogue.event.v1 \
+  Ce-Specversion:1.0 \
+  Ce-Application:app \
+  Ce-Device:device
+```
+
+And we can make sure that our changes took place:
+
+```shell
+http --auth ditto:ditto http://$TWIN_API/api/2/things/app_id:simple-thing
+HTTP/1.1 200 OK
+...
+etag: "rev:2"
+response-required: false
+version: 2
+
+{
+    "features": {
+        "hum": {
+            "properties": {
+                "value": "52.0"
+            }
+        },
+        "temp": {
+            "properties": {
+                "value": 22.5
+            }
+        }
+    },
+    "policyId": "app_id:simple-thing",
+    "thingId": "app_id:simple-thing"
+}
+```
+
+Why is this important? Well, it allow us to integrate it into the Knative workflows. Take a look at the following example:
+
+~~~yaml
+apiVersion: sources.knative.dev/v1alpha1
+kind: KafkaSource
+metadata:
+  name: ditto-converter-kafka-source
+  labels:
+    app.kubernetes.io/part-of: digital-twin
+spec:
+  consumerGroup: ditto-converter
+  bootstrapServers:
+    - kafka-eventing-kafka-bootstrap.knative-eventing.svc:9092
+  topics:
+    - knative-messaging-kafka.drogue-iot.iot-channel
+  sink:
+    ref:
+      apiVersion: flows.knative.dev/v1
+      kind: Sequence
+      name: ditto-converter
+---
+apiVersion: flows.knative.dev/v1
+kind: Sequence
+metadata:
+  name: ditto-converter
+  labels:
+    app.kubernetes.io/name: digital-twin
+    app.kubernetes.io/part-of: digital-twin
+spec:
+  channelTemplate:
+    apiVersion: messaging.knative.dev/v1alpha1
+    kind: KafkaChannel
+    spec:
+      numPartitions: 1
+      replicationFactor: 1
+  steps:
+    - ref:
+        apiVersion: serving.knative.dev/v1
+        kind: Service
+        name: drogue-ditto-converter
+  reply:
+    uri: http://ditto:ditto@ditto-nginx.drogue-iot.svc.cluster.local:8080/api/2/cloudevents
+~~~
+
+Here, we defined a `KafkaSource` with telemetry messages coming from our devices and wired it to the Knative `Sequence`. The sequence in this case will call our converter service and then post a result to Ditto's Cloud event API (with properly converted payload).
+
+OK, now it all makes more sense and we can start discussing how we are going to implement the converter service.
+
+## Func
+
+We already have a couple Knative services in Drogue cloud implemented with Rust and Quarkus, but they are hand-crafted from the ground up. Now, let's see how Func can help developing these kind of services.
+
+### Developing
+
+You can create a new project with:
+
+```shell
+func create --runtime quarkus --trigger events
+```
+
+This will create a Quarkus serverless project triggered by Cloud events instead of regular HTTP requests. Just what we need.
+
+You can find a full implementation of our service at [https://github.com/drogue-iot/drogue-ditto-converter], but let's focus here on some important snippets. The project will use [Quarkus Funqy support](https://quarkus.io/guides/funqy) to provide a stub of our service. Now we can implement it like:
```

Comment:
This link looks broken … You can omit the [] brackets if you want the plain HTTP link

Reply:
Fixed